### PR TITLE
Implement REPLCONF VERSION

### LINKED ISSUE #414: Replicas should indicate their version to their Primaries during handshake
I've seen it pop up in a few places where it be ideal to know what the version of the replica is we are replicating too, to potentially send something it might understand (and be potentially more compatible). I would suggest we extend the `REPLCONF` command to also support configuring the version metadata. Something like:
```
REPLCONF version 7.0.5
```

The replica would optionally send this after the other capabilities, so that it could gracefully handle the case where the primary does not support it. 

I thought about this after seeing, https://github.com/valkey-io/valkey/pull/245/files#r1586619641 (moved to follow-up issue #421), which would allow us to abort if a replica isn't on the same version as us.

### PR DESCRIPTION
**Pull request by zuiderkwast**
The replica sends when initiating replication, in pipeline with other REPLCONF commands.

The primary stores it in the client struct. Other fields are made smaller to avoid making the client struct consume more memory.

Fixes #414.

### CODE REVIEW

File: src/util.c [hunk 1]
```diff
@@ -872,6 +872,26 @@ int ld2string(char *buf, size_t len, long double value, ld2string_mode mode) {
...
+            part = 0;
+        } else {
+            return -1;
+        }
+        p++;
+    } while (*p);
+    v = (v << 8) | part;
+    return v;
+}
+
```
Comment (PingXie) [inline src/util.c]: I think we could tighten up the validation a bit so the DOS attach risk is addressed. 

1. the string length should be bounded, i.e., 12 chars at most
2. part should not exceed 0xff
3. there should be exactly 2 "."

File: src/util.c [hunk 2]
```diff
@@ -872,6 +872,26 @@ int ld2string(char *buf, size_t len, long double value, ld2string_mode mode) {
     return 0;
 }
 
+/* Parses a version string on the form "major.minor.patch" and returns an
+ * integer on the form 0xMMmmpp. Returns -1 on parse error. */
+int version2num(const char *version) {
```
Comment (PingXie) [inline src/util.c:877]:
```suggestion
int version2num(const sds version) {
```

**Review by PingXie — COMMENTED** [review]
LGTM overall

File: src/util.c (hunk 1 above)
Comment (zuiderkwast) [reply src/util.c]: I can't see how this is sensitive to any DOS attack. Invalid version numbers would just result in a version that's nonsense. The version parsing stops at the null terminator anyway. Do you think this computation is heavy for a long string? I think the network is the bottleneck in that case.

I can add the checks for other reasons though, like making it more strict to catch misbehaving clients. I think we shall document commands like REPLCONF better in the future to allow other programs/systems to connect as replicas.

File: src/util.c (hunk 2 above)
Comment (zuiderkwast) [reply src/util.c:877]: Why limit this to sds strings? It's unnecessary coupling IMO.

**Review by enjoy-binbin — APPROVED** [review]
LGTM.

File: src/util.c (hunk 2 above)
Comment (PingXie) [reply src/util.c:877]: sds has the length baked in so it is always more efficient to work with sds than c_str, in general

**Review by PingXie — APPROVED** [review]
LGTM

**Review by hwware — APPROVED** [review]
LGTM

File: src/util.c (hunk 2 above)
Comment (zuiderkwast) [reply src/util.c:877]: True, but util.c doesn't even include sds.h. All these functions are for `char *` (and sometimes a `size_t len` as a separate argument).

**Pull request merged**